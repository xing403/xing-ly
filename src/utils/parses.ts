import type { Array2DictOptions } from "../types";

/**
 * a array to dictionary
 * @param data
 * @param  config
 * @returns Object
 *
 * @case1
 *  [value1, value2] => [value1, value2]
 * @case2
 *  [{key: key1,name: name1}, {key: key2,name: name2}] => {key1: name1, key2:name2}
 */
export function Array2Dict<T extends Array<any>>(data: T, config: Array2DictOptions) {

  config = Object.assign({ deep: true, del_key: true, children: 'children' }, config)

  return data.reduce(function (obj, item) {
    const keys = Object.keys(item)
    if (!keys.includes(config.key))
      throw new Error(`the key \`${config.key}\` not exists`)

    if (['object', 'function', 'undefined', 'symbol'].includes(typeof item[config.key]))
      throw new Error(`the value of key \`${config.key}\` typeof \`${typeof item[config.key]}\`, is not acceptable type`)

    if (typeof item[config.key] === 'number' && Number.isNaN(item[config.key]))
      throw new Error(`the value of key \`${config.key}\` is NaN, is not acceptable type`)

   if (config.deep && config.children && item[config.children])
      item[config.children] = Array2Dict(item[config.children], config)

    if (keys.length == 1) { // do nothing => [item1, item2]
      return item

    } else if (keys.length == 2) { // transform [{key: value}]

      let value_key = keys.find(i => i != config.key) as string;
      obj[item[config.key]] = item[value_key];

    } else {  // transform [{ key: { key1: value1, key2: value2, ...} }]
      obj[item[config.key]] = item;
      if (config.del_key) delete item[config.key];  // delete the value corresponding to key
    }
    return obj;
  }, {});
}


