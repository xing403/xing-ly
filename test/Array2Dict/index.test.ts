import { Array2Dict } from '../../src/utils/parses'
import { expect, it } from 'vitest'
import ChinaCity from '../../src/packages/CitySelect/data'


it('Array2Dict', () => {
  expect(Array2Dict(ChinaCity[0], {
    key: 'name'
  })).toMatchSnapshot()
})

