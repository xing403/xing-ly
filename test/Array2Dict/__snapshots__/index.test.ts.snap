// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Array2Dict 1`] = `
{
  "上海市": {
    "children": {
      "嘉定区": "310114",
      "奉贤区": "310120",
      "宝山区": "310113",
      "崇明区": "310151",
      "徐汇区": "310104",
      "普陀区": "310107",
      "杨浦区": "310110",
      "松江区": "310117",
      "浦东新区": "310115",
      "虹口区": "310109",
      "金山区": "310116",
      "长宁区": "310105",
      "闵行区": "310112",
      "青浦区": "310118",
      "静安区": "310106",
      "黄浦区": "310101",
    },
    "code": "31",
  },
  "云南省": {
    "children": {
      "临沧市": "5309",
      "丽江市": "5307",
      "保山市": "5305",
      "大理白族自治州": "5329",
      "德宏傣族景颇族自治州": "5331",
      "怒江傈僳族自治州": "5333",
      "文山壮族苗族自治州": "5326",
      "昆明市": "5301",
      "昭通市": "5306",
      "普洱市": "5308",
      "曲靖市": "5303",
      "楚雄彝族自治州": "5323",
      "玉溪市": "5304",
      "红河哈尼族彝族自治州": "5325",
      "西双版纳傣族自治州": "5328",
      "迪庆藏族自治州": "5334",
    },
    "code": "53",
  },
  "内蒙古自治区": {
    "children": {
      "乌兰察布市": "1509",
      "乌海市": "1503",
      "兴安盟": "1522",
      "包头市": "1502",
      "呼伦贝尔市": "1507",
      "呼和浩特市": "1501",
      "巴彦淖尔市": "1508",
      "赤峰市": "1504",
      "通辽市": "1505",
      "鄂尔多斯市": "1506",
      "锡林郭勒盟": "1525",
      "阿拉善盟": "1529",
    },
    "code": "15",
  },
  "北京市": {
    "children": {
      "东城区": "110101",
      "丰台区": "110106",
      "大兴区": "110115",
      "密云区": "110118",
      "平谷区": "110117",
      "延庆区": "110119",
      "怀柔区": "110116",
      "房山区": "110111",
      "昌平区": "110114",
      "朝阳区": "110105",
      "海淀区": "110108",
      "石景山区": "110107",
      "西城区": "110102",
      "通州区": "110112",
      "门头沟区": "110109",
      "顺义区": "110113",
    },
    "code": "11",
  },
  "吉林省": {
    "children": {
      "吉林市": "2202",
      "四平市": "2203",
      "延边朝鲜族自治州": "2224",
      "松原市": "2207",
      "白城市": "2208",
      "白山市": "2206",
      "辽源市": "2204",
      "通化市": "2205",
      "长春市": "2201",
    },
    "code": "22",
  },
  "四川省": {
    "children": {
      "乐山市": "5111",
      "内江市": "5110",
      "凉山彝族自治州": "5134",
      "南充市": "5113",
      "宜宾市": "5115",
      "巴中市": "5119",
      "广元市": "5108",
      "广安市": "5116",
      "德阳市": "5106",
      "成都市": "5101",
      "攀枝花市": "5104",
      "泸州市": "5105",
      "甘孜藏族自治州": "5133",
      "眉山市": "5114",
      "绵阳市": "5107",
      "自贡市": "5103",
      "资阳市": "5120",
      "达州市": "5117",
      "遂宁市": "5109",
      "阿坝藏族羌族自治州": "5132",
      "雅安市": "5118",
    },
    "code": "51",
  },
  "天津市": {
    "children": {
      "东丽区": "120110",
      "北辰区": "120113",
      "南开区": "120104",
      "和平区": "120101",
      "宁河区": "120117",
      "宝坻区": "120115",
      "武清区": "120114",
      "河东区": "120102",
      "河北区": "120105",
      "河西区": "120103",
      "津南区": "120112",
      "滨海新区": "120116",
      "红桥区": "120106",
      "蓟州区": "120119",
      "西青区": "120111",
      "静海区": "120118",
    },
    "code": "12",
  },
  "宁夏回族自治区": {
    "children": {
      "中卫市": "6405",
      "吴忠市": "6403",
      "固原市": "6404",
      "石嘴山市": "6402",
      "银川市": "6401",
    },
    "code": "64",
  },
  "安徽省": {
    "children": {
      "亳州市": "3416",
      "六安市": "3415",
      "合肥市": "3401",
      "安庆市": "3408",
      "宣城市": "3418",
      "宿州市": "3413",
      "池州市": "3417",
      "淮北市": "3406",
      "淮南市": "3404",
      "滁州市": "3411",
      "芜湖市": "3402",
      "蚌埠市": "3403",
      "铜陵市": "3407",
      "阜阳市": "3412",
      "马鞍山市": "3405",
      "黄山市": "3410",
    },
    "code": "34",
  },
  "山东省": {
    "children": {
      "东营市": "3705",
      "临沂市": "3713",
      "威海市": "3710",
      "德州市": "3714",
      "日照市": "3711",
      "枣庄市": "3704",
      "泰安市": "3709",
      "济南市": "3701",
      "济宁市": "3708",
      "淄博市": "3703",
      "滨州市": "3716",
      "潍坊市": "3707",
      "烟台市": "3706",
      "聊城市": "3715",
      "菏泽市": "3717",
      "青岛市": "3702",
    },
    "code": "37",
  },
  "山西省": {
    "children": {
      "临汾市": "1410",
      "吕梁市": "1411",
      "大同市": "1402",
      "太原市": "1401",
      "忻州市": "1409",
      "晋中市": "1407",
      "晋城市": "1405",
      "朔州市": "1406",
      "运城市": "1408",
      "长治市": "1404",
      "阳泉市": "1403",
    },
    "code": "14",
  },
  "广东省": {
    "children": {
      "东莞市": "4419",
      "中山市": "4420",
      "云浮市": "4453",
      "佛山市": "4406",
      "广州市": "4401",
      "惠州市": "4413",
      "揭阳市": "4452",
      "梅州市": "4414",
      "汕头市": "4405",
      "汕尾市": "4415",
      "江门市": "4407",
      "河源市": "4416",
      "深圳市": "4403",
      "清远市": "4418",
      "湛江市": "4408",
      "潮州市": "4451",
      "珠海市": "4404",
      "肇庆市": "4412",
      "茂名市": "4409",
      "阳江市": "4417",
      "韶关市": "4402",
    },
    "code": "44",
  },
  "广西壮族自治区": {
    "children": {
      "北海市": "4505",
      "南宁市": "4501",
      "崇左市": "4514",
      "来宾市": "4513",
      "柳州市": "4502",
      "桂林市": "4503",
      "梧州市": "4504",
      "河池市": "4512",
      "玉林市": "4509",
      "百色市": "4510",
      "贵港市": "4508",
      "贺州市": "4511",
      "钦州市": "4507",
      "防城港市": "4506",
    },
    "code": "45",
  },
  "新疆维吾尔自治区": {
    "children": {
      "乌鲁木齐市": "6501",
      "五家渠市": "659004",
      "伊犁哈萨克自治州": "6540",
      "克孜勒苏柯尔克孜自治州": "6530",
      "克拉玛依市": "6502",
      "北屯市": "659005",
      "博尔塔拉蒙古自治州": "6527",
      "双河市": "659007",
      "可克达拉市": "659008",
      "吐鲁番市": "6504",
      "和田地区": "6532",
      "哈密市": "6505",
      "喀什地区": "6531",
      "图木舒克市": "659003",
      "塔城地区": "6542",
      "巴音郭楞蒙古自治州": "6528",
      "新星市": "659011",
      "昆玉市": "659009",
      "昌吉回族自治州": "6523",
      "白杨市": "659012",
      "石河子市": "659001",
      "胡杨河市": "659010",
      "铁门关市": "659006",
      "阿克苏地区": "6529",
      "阿勒泰地区": "6543",
      "阿拉尔市": "659002",
    },
    "code": "65",
  },
  "江苏省": {
    "children": {
      "南京市": "3201",
      "南通市": "3206",
      "宿迁市": "3213",
      "常州市": "3204",
      "徐州市": "3203",
      "扬州市": "3210",
      "无锡市": "3202",
      "泰州市": "3212",
      "淮安市": "3208",
      "盐城市": "3209",
      "苏州市": "3205",
      "连云港市": "3207",
      "镇江市": "3211",
    },
    "code": "32",
  },
  "江西省": {
    "children": {
      "上饶市": "3611",
      "九江市": "3604",
      "南昌市": "3601",
      "吉安市": "3608",
      "宜春市": "3609",
      "抚州市": "3610",
      "新余市": "3605",
      "景德镇市": "3602",
      "萍乡市": "3603",
      "赣州市": "3607",
      "鹰潭市": "3606",
    },
    "code": "36",
  },
  "河北省": {
    "children": {
      "保定市": "1306",
      "唐山市": "1302",
      "廊坊市": "1310",
      "张家口市": "1307",
      "承德市": "1308",
      "沧州市": "1309",
      "石家庄市": "1301",
      "秦皇岛市": "1303",
      "衡水市": "1311",
      "邢台市": "1305",
      "邯郸市": "1304",
    },
    "code": "13",
  },
  "河南省": {
    "children": {
      "三门峡市": "4112",
      "信阳市": "4115",
      "南阳市": "4113",
      "周口市": "4116",
      "商丘市": "4114",
      "安阳市": "4105",
      "平顶山市": "4104",
      "开封市": "4102",
      "新乡市": "4107",
      "洛阳市": "4103",
      "济源市": "419001",
      "漯河市": "4111",
      "濮阳市": "4109",
      "焦作市": "4108",
      "许昌市": "4110",
      "郑州市": "4101",
      "驻马店市": "4117",
      "鹤壁市": "4106",
    },
    "code": "41",
  },
  "浙江省": {
    "children": {
      "丽水市": "3311",
      "台州市": "3310",
      "嘉兴市": "3304",
      "宁波市": "3302",
      "杭州市": "3301",
      "温州市": "3303",
      "湖州市": "3305",
      "绍兴市": "3306",
      "舟山市": "3309",
      "衢州市": "3308",
      "金华市": "3307",
    },
    "code": "33",
  },
  "海南省": {
    "children": {
      "万宁市": "469006",
      "三亚市": "4602",
      "三沙市": "4603",
      "东方市": "469007",
      "临高县": "469024",
      "乐东黎族自治县": "469027",
      "五指山市": "469001",
      "保亭黎族苗族自治县": "469029",
      "儋州市": "4604",
      "定安县": "469021",
      "屯昌县": "469022",
      "文昌市": "469005",
      "昌江黎族自治县": "469026",
      "海口市": "4601",
      "澄迈县": "469023",
      "琼中黎族苗族自治县": "469030",
      "琼海市": "469002",
      "白沙黎族自治县": "469025",
      "陵水黎族自治县": "469028",
    },
    "code": "46",
  },
  "湖北省": {
    "children": {
      "仙桃市": "429004",
      "十堰市": "4203",
      "咸宁市": "4212",
      "天门市": "429006",
      "孝感市": "4209",
      "宜昌市": "4205",
      "恩施土家族苗族自治州": "4228",
      "武汉市": "4201",
      "潜江市": "429005",
      "神农架林区": "429021",
      "荆州市": "4210",
      "荆门市": "4208",
      "襄阳市": "4206",
      "鄂州市": "4207",
      "随州市": "4213",
      "黄冈市": "4211",
      "黄石市": "4202",
    },
    "code": "42",
  },
  "湖南省": {
    "children": {
      "娄底市": "4313",
      "岳阳市": "4306",
      "常德市": "4307",
      "张家界市": "4308",
      "怀化市": "4312",
      "株洲市": "4302",
      "永州市": "4311",
      "湘潭市": "4303",
      "湘西土家族苗族自治州": "4331",
      "益阳市": "4309",
      "衡阳市": "4304",
      "邵阳市": "4305",
      "郴州市": "4310",
      "长沙市": "4301",
    },
    "code": "43",
  },
  "甘肃省": {
    "children": {
      "临夏回族自治州": "6229",
      "兰州市": "6201",
      "嘉峪关市": "6202",
      "天水市": "6205",
      "定西市": "6211",
      "平凉市": "6208",
      "庆阳市": "6210",
      "张掖市": "6207",
      "武威市": "6206",
      "甘南藏族自治州": "6230",
      "白银市": "6204",
      "酒泉市": "6209",
      "金昌市": "6203",
      "陇南市": "6212",
    },
    "code": "62",
  },
  "福建省": {
    "children": {
      "三明市": "3504",
      "南平市": "3507",
      "厦门市": "3502",
      "宁德市": "3509",
      "泉州市": "3505",
      "漳州市": "3506",
      "福州市": "3501",
      "莆田市": "3503",
      "龙岩市": "3508",
    },
    "code": "35",
  },
  "西藏自治区": {
    "children": {
      "山南市": "5405",
      "拉萨市": "5401",
      "日喀则市": "5402",
      "昌都市": "5403",
      "林芝市": "5404",
      "那曲市": "5406",
      "阿里地区": "5425",
    },
    "code": "54",
  },
  "贵州省": {
    "children": {
      "六盘水市": "5202",
      "安顺市": "5204",
      "毕节市": "5205",
      "贵阳市": "5201",
      "遵义市": "5203",
      "铜仁市": "5206",
      "黔东南苗族侗族自治州": "5226",
      "黔南布依族苗族自治州": "5227",
      "黔西南布依族苗族自治州": "5223",
    },
    "code": "52",
  },
  "辽宁省": {
    "children": {
      "丹东市": "2106",
      "大连市": "2102",
      "抚顺市": "2104",
      "朝阳市": "2113",
      "本溪市": "2105",
      "沈阳市": "2101",
      "盘锦市": "2111",
      "营口市": "2108",
      "葫芦岛市": "2114",
      "辽阳市": "2110",
      "铁岭市": "2112",
      "锦州市": "2107",
      "阜新市": "2109",
      "鞍山市": "2103",
    },
    "code": "21",
  },
  "重庆市": {
    "children": {
      "万州区": "500101",
      "丰都县": "500230",
      "九龙坡区": "500107",
      "云阳县": "500235",
      "北碚区": "500109",
      "南岸区": "500108",
      "南川区": "500119",
      "合川区": "500117",
      "垫江县": "500231",
      "城口县": "500229",
      "大渡口区": "500104",
      "大足区": "500111",
      "奉节县": "500236",
      "巫山县": "500237",
      "巫溪县": "500238",
      "巴南区": "500113",
      "开州区": "500154",
      "彭水苗族土家族自治县": "500243",
      "忠县": "500233",
      "梁平区": "500155",
      "武隆区": "500156",
      "永川区": "500118",
      "江北区": "500105",
      "江津区": "500116",
      "沙坪坝区": "500106",
      "涪陵区": "500102",
      "渝中区": "500103",
      "渝北区": "500112",
      "潼南区": "500152",
      "璧山区": "500120",
      "石柱土家族自治县": "500240",
      "秀山土家族苗族自治县": "500241",
      "綦江区": "500110",
      "荣昌区": "500153",
      "酉阳土家族苗族自治县": "500242",
      "铜梁区": "500151",
      "长寿区": "500115",
      "黔江区": "500114",
    },
    "code": "50",
  },
  "陕西省": {
    "children": {
      "咸阳市": "6104",
      "商洛市": "6110",
      "安康市": "6109",
      "宝鸡市": "6103",
      "延安市": "6106",
      "榆林市": "6108",
      "汉中市": "6107",
      "渭南市": "6105",
      "西安市": "6101",
      "铜川市": "6102",
    },
    "code": "61",
  },
  "青海省": {
    "children": {
      "果洛藏族自治州": "6326",
      "海东市": "6302",
      "海北藏族自治州": "6322",
      "海南藏族自治州": "6325",
      "海西蒙古族藏族自治州": "6328",
      "玉树藏族自治州": "6327",
      "西宁市": "6301",
      "黄南藏族自治州": "6323",
    },
    "code": "63",
  },
  "黑龙江省": {
    "children": {
      "七台河市": "2309",
      "伊春市": "2307",
      "佳木斯市": "2308",
      "双鸭山市": "2305",
      "哈尔滨市": "2301",
      "大兴安岭地区": "2327",
      "大庆市": "2306",
      "牡丹江市": "2310",
      "绥化市": "2312",
      "鸡西市": "2303",
      "鹤岗市": "2304",
      "黑河市": "2311",
      "齐齐哈尔市": "2302",
    },
    "code": "23",
  },
}
`;
